Source/qol/autopickup.cpp

```diff
@@ -30,31 +31,73 @@ bool HasRoomForGold()
 	return false;
 }
 
+int numMiscItemsInInv(int iMiscId)
+{
+	int numItems = 0;
+	for (const auto &item : InventoryAndBeltPlayerItemsRange(Players[MyPlayerId])) {
+		if (item._iMiscId == iMiscId) {
+			numItems++;
+		}
+	}
+	return numItems;
+}
+
 } // namespace
 
-void AutoGoldPickup(int pnum)
+void AutoPickup(int pnum)
 {
-	if (!sgOptions.Gameplay.bAutoGoldPickup)
+	if (leveltype == DTYPE_TOWN && !sgOptions.Gameplay.bAutoPickupInTown)
 		return;
-
 	if (pnum != MyPlayerId)
 		return;
-	if (leveltype == DTYPE_TOWN)
-		return;
-	if (!HasRoomForGold())
-		return;
+
+	bool hasRoomForGold = HasRoomForGold();
 
 	for (auto pathDir : PathDirs) {
 		Point tile = Players[pnum].position.tile + pathDir;
 		if (dItem[tile.x][tile.y] != 0) {
 			int itemIndex = dItem[tile.x][tile.y] - 1;
 			auto &item = Items[itemIndex];
-			if (item._itype == ItemType::Gold) {
+			if (hasRoomForGold && item._itype == ItemType::Gold) {
 				NetSendCmdGItem(true, CMD_REQUESTAGITEM, pnum, pnum, itemIndex);
 				item._iRequest = true;
 			}
+			if (AutoPlaceItemInInventory(Players[pnum], item, false) || AutoPlaceItemInBelt(Players[pnum], item, false)) {

```

Comment:
It'd be worth having a check for item._itype == ItemType::Misc here. It's mainly for clarity but also avoids running the inventory size calcs for items that will never be picked up automatically. _iMiscId is occasionally populated for non-misc items, though this currently doesn't impact the functionality as potion/elixir subtypes aren't reused.
```suggestion
			if (item._itype == ItemType::Misc && (AutoPlaceItemInInventory(Players[pnum], item, false) || AutoPlaceItemInBelt(Players[pnum], item, false))) {
```